Model/src/main/java/org/apidb/apicommon/model/report/bed/feature/ProteinSequenceFeatureProvider.java

```diff
@@ -0,0 +1,143 @@
+import org.gusdb.wdk.model.WdkModelException;
+import org.gusdb.wdk.model.record.RecordInstance;
+import org.json.JSONObject;
+import org.apidb.apicommon.model.report.bed.util.StrandDirection;
+import org.apidb.apicommon.model.report.bed.util.RequestedDeflineFields;
+import org.apidb.apicommon.model.report.bed.util.DeflineBuilder;
+import org.apidb.apicommon.model.report.bed.util.BedLine;
+
+public class ProteinSequenceFeatureProvider implements BedFeatureProvider {
+
+  private enum ProteinAnchor {
+    DownstreamFromStart,
+    UpstreamFromEnd
+  }
+
+  private static final String ATTR_PROTEIN_LENGTH = "protein_length";
+  private static final String ATTR_ORGANISM = "organism";
+  private static final String ATTR_GENE_NAME = "gene_name";
+  private static final String ATTR_GENE_PRODUCT = "gene_product";
+
+  private final RequestedDeflineFields _requestedDeflineFields;
+  private final int _startOffset;
+  private final ProteinAnchor _startAnchor;
+  private final int _endOffset;
+  private final ProteinAnchor _endAnchor;
+
+  public ProteinSequenceFeatureProvider(JSONObject config) {
+    _requestedDeflineFields = new RequestedDeflineFields(config);
+
+    _startOffset = config.getInt("startOffset3");
+    _startAnchor = ProteinAnchor.valueOf(config.getString("startAnchor3"));
+
+    _endOffset = config.getInt("endOffset3");
+    _endAnchor = ProteinAnchor.valueOf(config.getString("endAnchor3"));
+  }
+
+  @Override
+  public String getRequiredRecordClassFullName() {
+    return TranscriptUtil.TRANSCRIPT_RECORDCLASS;
+  }
+
+  @Override
+  public String[] getRequiredAttributeNames() {
+    return new String[] {
+        ATTR_PROTEIN_LENGTH,
+        ATTR_ORGANISM,
+        ATTR_GENE_NAME,
+        ATTR_GENE_PRODUCT
+    };
+  }
+
+  @Override
+  public String[] getRequiredTableNames() {
+    return new String[0];
+  }
+
+  @Override
+  public List<List<String>> getRecordAsBedFields(RecordInstance record) throws WdkModelException {
+    /*
+     * Not all genes have a protein sequence - only protein coding genes do
+     * Use emptiness of ATTR_PROTEIN_LENGTH as sentinel value
+     */
+    String featureLengthStringOrEmpty = stringValue(record, ATTR_PROTEIN_LENGTH);
+    if("".equals(featureLengthStringOrEmpty)){
+      return List.of();
+    } else {
+      return List.of(proteinBedLineForProteinCodingGene(record, Integer.valueOf(featureLengthStringOrEmpty)));
+    }
+  }
+
+  public List<String> proteinBedLineForProteinCodingGene(RecordInstance record, Integer featureLength) throws WdkModelException {
+    String featureId = getSourceId(record);
+    String chrom = featureId;
+    StrandDirection strand = StrandDirection.none;
+
+    Integer segmentStart = getPositionProtein(featureLength, _startOffset, _startAnchor);
+    Integer segmentEnd = getPositionProtein(featureLength, _endOffset, _endAnchor);
+
+    DeflineBuilder defline = new DeflineBuilder(featureId);
+
+    if(_requestedDeflineFields.contains("organism")){
```

Comment:
I've seen this pattern a bunch of times.  I wonder if it could be easily factored into something that knows how to provide the parts of a pretty defline for the various types?

Reply:
Probably more trouble than it's worth.